File: demos_code_J3_conception_objet/07_mde_mini_generateur/src/main/java/MiniMdeGenerator.java
import java.nio.file.*;
import java.util.*;

public final class MiniMdeGenerator {
    private MiniMdeGenerator() {}

    public static Path generate(Path model, Path outputDir) throws Exception {
        List<String> lines = Files.readAllLines(model);
        if (lines.isEmpty() || !lines.get(0).startsWith("entity ")) {
            throw new IllegalArgumentException("modèle invalide");
        }

        String name = lines.get(0).substring("entity ".length()).trim();
        List<String> fields = new ArrayList<>();
        for (int i = 1; i < lines.size(); i++) {
            String[] p = lines.get(i).trim().split("\s+");
            if (p.length == 3 && p[0].equals("field")) {
                fields.add(p[2] + " " + p[1]);
            }
        }

        String code = "public record " + name + "(" + String.join(", ", fields) + ") {}\n";
        Files.createDirectories(outputDir);
        Path generated = outputDir.resolve(name + ".java");
        Files.writeString(generated, code);
        return generated;
    }
}

File: Demo en Python/demos_code_J3_conception_objet_python/07_mde_mini_generateur/mini_mde_generator.py
from pathlib import Path

TYPE_MAP = {
    "String": "str",  # accepte aussi le vocabulaire de la démo Java
    "str": "str",
    "int": "int",
    "boolean": "bool",
    "bool": "bool",
    "double": "float",
    "float": "float",
}


def generate(model: Path, output_dir: Path) -> Path:
    lines = model.read_text(encoding="utf-8").splitlines()
    if not lines or not lines[0].startswith("entity "):
        raise ValueError("modèle invalide")

    name = lines[0][len("entity "):].strip()
    fields = []
    for line in lines[1:]:
        parts = line.strip().split()
        if len(parts) == 3 and parts[0] == "field":
            field_name, field_type = parts[1], parts[2]
            fields.append((field_name, TYPE_MAP.get(field_type, field_type)))

    body = ["from dataclasses import dataclass", "", "", "@dataclass(frozen=True)", f"class {name}:"]
    if fields:
        body.extend(f"    {field_name}: {field_type}" for field_name, field_type in fields)
    else:
        body.append("    pass")
    code = "\n".join(body) + "\n"

    output_dir.mkdir(parents=True, exist_ok=True)
    generated = output_dir / f"{name.lower()}.py"
    generated.write_text(code, encoding="utf-8")
    return generated
